### PR TITLE
First initial implementation with azure provider

### PR DESCRIPTION
**Pull request by splattner**
This first implementation only works for the azure provider and is based on https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/azure.md

## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Update the documentation.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog

### CODE REVIEW

File: component/main.jsonnet [hunk 1]
```diff
@@ -5,6 +5,120 @@ local inv = kap.inventory();
...
+  subjects_: [ serviceaccount ],
+};
+
+// https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/azure.md#creating-configuration-file
+local azuresecret = kube.Secret('azure-config-file') {
+  metadata+: {
+    namespace: params.namespace,
+  },
+  stringData+: {
+    'azure.json': '{\n      tenantId: ' + params.providerConfig.azure.authentication.tenantId + ',\n      subscriptionId: ' + params.providerConfig.azure.authentication.subscriptionId + ',\n      resourceGroup: ' + params.providerConfig.azure.authentication.resourceGroup + ',\n      aadClientId: ' + params.providerConfig.azure.authentication.aadClientId + ',\n      aadClientSecret: ' + params.providerConfig.azure.authentication.aadClientSecret + '\n    }',
```
Comment (splattner) [inline component/main.jsonnet]: not sure if this can be done a bit nicer?

File: class/defaults.yml [hunk 2]
```diff
@@ -2,3 +2,27 @@ parameters:
   external_dns:
     =_metadata: {}
     namespace: syn-external-dns
+
+    images:
+      externaldns:
+        registry: k8s.gcr.io
+        image: external-dns
+        tag: v0.8.0
+
```
Comment (megian) [inline class/defaults.yml:11]: Not required newline

File: class/defaults.yml [hunk 3]
```diff
@@ -2,3 +2,27 @@ parameters:
...
+
+    provider: azure
```
Comment (megian) [inline class/defaults.yml]:
```suggestion
    provider: ${facts:cloud}
```

File: class/defaults.yml [hunk 4]
```diff
@@ -2,3 +2,27 @@ parameters:
...
+
+    providerConfig:
+      azure:
```
Comment (megian) [inline class/defaults.yml:15]: Either using a map like https://github.com/projectsyn/component-keycloak/blob/master/class/defaults.yml#L9 for default values. Or moving that to the documentation -> how-to.

File: component/main.jsonnet [hunk 5]
```diff
@@ -5,6 +5,120 @@ local inv = kap.inventory();
...
 local params = inv.parameters.external_dns;
 
+local namespace = kube.Namespace(params.namespace);
+
+local serviceaccount = kube.ServiceAccount('external-dns') {
+  metadata+: {
+    namespace: params.namespace,
+  },
+};
+
```
Comment (megian) [inline component/main.jsonnet:15]: Not required newline

File: component/main.jsonnet [hunk 6]
```diff
@@ -5,6 +5,120 @@ local inv = kap.inventory();
...
+      verbs: [ 'get', 'watch', 'list' ],
+    },
+    {
+      apiGroups: [ '' ],
+      resources: [ 'nodes' ],
+      verbs: [ 'list' ],
+    },
+  ],
+};
+
```
Comment (megian) [inline component/main.jsonnet:35]: Not required newline

File: component/main.jsonnet [hunk 7]
```diff
@@ -5,6 +5,120 @@ local inv = kap.inventory();
...
+// https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/azure.md#creating-configuration-file
+local azuresecret = kube.Secret('azure-config-file') {
+  metadata+: {
+    namespace: params.namespace,
+  },
+  stringData+: {
+    'azure.json': '{\n      tenantId: ' + params.providerConfig.azure.authentication.tenantId + ',\n      subscriptionId: ' + params.providerConfig.azure.authentication.subscriptionId + ',\n      resourceGroup: ' + params.providerConfig.azure.authentication.resourceGroup + ',\n      aadClientId: ' + params.providerConfig.azure.authentication.aadClientId + ',\n      aadClientSecret: ' + params.providerConfig.azure.authentication.aadClientSecret + '\n    }',
+  },
+};
+
```
Comment (megian) [inline component/main.jsonnet:52]: Not required newline

File: component/main.jsonnet (hunk 1 above)
Comment (megian) [reply component/main.jsonnet]: Not sure if `std.manifestJsonEx(value, indent)` https://jsonnet.org/ref/stdlib.html might work.

File: component/main.jsonnet [hunk 8]
```diff
@@ -5,6 +5,120 @@ local inv = kap.inventory();
...
+
+local mountVolumes =
+  if params.provider == 'azure' then
+    [
+      {
+        name: 'azure-config-file',
```
Comment (megian) [inline component/main.jsonnet:57]: Is it required to have the provider name in. Could this be simplified using a generic name?

File: component/main.jsonnet [hunk 9]
```diff
@@ -5,6 +5,120 @@ local inv = kap.inventory();
...
+        readOnly: true,
+      },
+    ]
+  else [];
+
+local volumes =
+  if params.provider == 'azure' then
+    [
+      {
+        name: 'azure-config-file',
```
Comment (megian) [inline component/main.jsonnet:68]: ditto?

File: component/main.jsonnet [hunk 10]
```diff
@@ -5,6 +5,120 @@ local inv = kap.inventory();
...
+            {
+              key: 'externaldns-config.json',
+              path: 'azure.json',
+            },
+          ],
+        },
+      },
+    ]
+  else [];
+
```
Comment (megian) [inline component/main.jsonnet:81]: ditto?

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 11]
```diff
@@ -10,6 +10,45 @@ default:: `syn-external-dns`
 
 The namespace in which to deploy this component.
 
+== `provider`
+
+[horizontal]
+type:: string
+default:: `azure`
+
+The Provider to use
```
Comment (megian) [inline docs/modules/ROOT/pages/references/parameters.adoc]:
```suggestion
The Provider to use.
```

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 12]
```diff
@@ -10,6 +10,45 @@ default:: `syn-external-dns`
 
 The namespace in which to deploy this component.
 
+== `provider`
+
+[horizontal]
+type:: string
+default:: `azure`
```
Comment (megian) [inline docs/modules/ROOT/pages/references/parameters.adoc]:
```suggestion
default:: `${facts:cloud}`
```

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 13]
```diff
@@ -10,6 +10,45 @@ default:: `syn-external-dns`
...
+
+The Provider to use
+
+== `providerConfig.azure`
+
+[horizontal]
+type:: dict
+default:: `azure`
+
+Azure Provider specific configuration. See defaults.yml
```
Comment (megian) [inline docs/modules/ROOT/pages/references/parameters.adoc]: Usually we adding a how-to document with an intro how to configure it. https://github.com/projectsyn/component-keycloak/blob/master/docs/modules/ROOT/pages/how-tos/configure-ingress.adoc

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 14]
```diff
@@ -10,6 +10,45 @@ default:: `syn-external-dns`
...
+
+== `config.source`
+
+[horizontal]
+type:: string
+default:: `ingress`
+
+The source from where external-dns gets the records to insert into the DNS Zone
```
Comment (megian) [inline docs/modules/ROOT/pages/references/parameters.adoc]:
```suggestion
The source from where external-dns gets the records to insert into the DNS zone.
```

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 15]
```diff
@@ -10,6 +10,45 @@ default:: `syn-external-dns`
...
+
+== `config.txtPrefix`
+
+[horizontal]
+type:: string
+default:: `externaldns-`
+
+Prefix for txt records which are used by external-dns for marking ownership of DNS Records
```
Comment (megian) [inline docs/modules/ROOT/pages/references/parameters.adoc]:
```suggestion
Prefix for txt records which are used by external-dns for marking ownership of DNS records.
```

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 16]
```diff
@@ -10,6 +10,45 @@ default:: `syn-external-dns`
...
+
+== `config.domainFilter`
+
+[horizontal]
+type:: string
+default:: ``
+
+limit to only the selected domains
```
Comment (megian) [inline docs/modules/ROOT/pages/references/parameters.adoc]:
```suggestion
Limit to only the selected domains.
```

File: class/defaults.yml (hunk 3 above)
Comment (splattner) [reply class/defaults.yml]: this time its not really linked to the cloud provider but to one one these: https://github.com/kubernetes-sigs/external-dns#the-latest-release-v08
therefore I think its better to not use the cloud fact

File: component/main.jsonnet (hunk 1 above)
Comment (splattner) [reply component/main.jsonnet]: yeah, thats better, thx!

File: component/main.jsonnet (hunk 8 above)
Comment (splattner) [reply component/main.jsonnet:57]: not really, but each provider does it a bit different. I'll check if/how I could make this a bit more generic

Comment (splattner) [reply component/main.jsonnet:57]: yeah, some are using env, some volumes, some arguments... I wan't this to be the same an in the provider documentation. Once more provider are implemented, we can think of a better solution.

File: class/defaults.yml (hunk 4 above)
Comment (splattner) [reply class/defaults.yml:15]: I documented it in the parameter section and also gonna write an how-to

File: component/main.jsonnet (hunk 10 above)
Comment (splattner) [reply component/main.jsonnet:81]: see above, unfortunate each provider does it a bit different. 
Currently the implementation is for azure, once we need more external-dns provider, we should look into this again and see if a more generic way is possible.

**Pull request merged**